haigha/classes/basic_class.py

```diff
@@ -53,8 +54,30 @@ def _cleanup(self):
         self._get_cb = None
         self._recover_cb = None
         self._cancel_cb = None
+        self._return_listener = None
```

Comment:
@awestendorf, I have an idea that can work well for the publisher-acks use case. However, I don't think that we can make things better for the non-confirmation scenario, because there is no Basic.Ack/Nack in that case, so the time to receive a basic.return is unbounded.

Here is the background: Referring to the discussion in https://groups.google.com/forum/#!topic/rabbitmq-users/RRSFgt2A9Kw, Michael Klishin (Pivotal/RabbitMQ) is contemplating adding an extended property to the header table of the Header Frame of basic.return methods; this extended property would contain the delivery-tag value of the returned message that would permit client libraries to match it against the expected delivery tags of messages that they published. Until then, Haigha may implement the equivalent on its own by adding its own extended property with the expected delivery-tag value, and then looking it up when handling basic.return. So, depending on your perspective, haigha can either wait for the standard mechanism to come online or implement a simple equivalent that it would use with broker versions that don't have built-in support. I favor the latter.

So, this is what I was thinking for the publisher-acks use case:

1. User enables pubacks
2. When calling `channel.basic.publish()`, the user passes a `puback_cb(method, returned_msg)` (conceptual, callback signature TBD)
3. Haigha saves the delivery-tag and callback in the channel's (new) pubacks dictionary
4. On receipt of Basic.Return, the channel saves the necessary returned info, including the returned message, in the matching property in the pubacks dict, alongside the user's callback.
5. On receipt of Basic.Ack/Nack (it contains the delivery-tag in the method frame), the channel removes the matching property from the pubacks dict and calls the user's callback with the necessary info, including the returned message, if any. This works out, because \- in the pubacks scenario \- Basic.Return is guaranteed to arrive before Basic.Ack.